### PR TITLE
Implement MSC3816, consider the root event for thread participation

### PR DESCRIPTION
**Pull request by clokep**
Implement support for [MSC3816](https://github.com/matrix-org/matrix-spec-proposals/pull/3816) to consider having the "root" event in a thread as participated in the thread.

### CODE REVIEW

File: synapse/handlers/relations.py [hunk 1]
```diff
@@ -270,16 +267,34 @@ async def get_threads_for_events(
...
+        # Only fetch whether the current user has participated in a thread based
+        # on which returned for a limited selection based on what had
         # summaries.
         thread_event_ids = [
             event_id for event_id, summary in summaries.items() if summary
         ]
-        participated = await self._main_store.get_threads_participated(
-            thread_event_ids, user_id
+
+        # Preseed thread participation with whether the requester send the event.
```
Comment (MadLittleMods) [inline synapse/handlers/relations.py:283]:
```suggestion
        # Pre-seed thread participation with whether the requester sent the event.
```

Comment (MadLittleMods) [reply synapse/handlers/relations.py:283]: Maybe if these are always root thread events:

```suggestion
        # Pre-seed thread participation with whether the requester sent the root event.
```

File: synapse/handlers/relations.py [hunk 2]
```diff
@@ -270,16 +267,34 @@ async def get_threads_for_events(
...
+        participated = {
+            event_id: events_by_id[event_id].sender == user_id
+            for event_id in thread_event_ids
+        }
+        # Check other events against the database.
```
Comment (MadLittleMods) [inline synapse/handlers/relations.py:288]:
```suggestion
        # Check thread participation (the rest of the events in the thread) against the database
```

File: tests/rest/client/test_relations.py [hunk 3]
```diff
@@ -1037,30 +1040,52 @@ def test_thread(self) -> None:
...
-                bundled_aggregations.get("latest_event"),
-            )
+                    bundled_aggregations.get("latest_event"),
+                )
+
+            return assert_thread
 
-        self._test_bundled_aggregations(RelationTypes.THREAD, assert_thread, 10)
+        # A user which has sent the root event or replied has participated.
+        self._test_bundled_aggregations(RelationTypes.THREAD, _gen_assert(True), 9)
```
Comment (MadLittleMods) [inline tests/rest/client/test_relations.py:1076]: The shared bits of these tests make it pretty hard to follow. "user 1" participated because `_test_bundled_aggregations` uses `self.parent_id` (the thread root event) by default which was sent by `self.user_token` in `prepare`.

This test would be a lot more clear if we passed in the root event ID and token explicitly.

We could call the token `user_who_sent_parent_event_token` for the person who sent `parent_event_id` to make an even stronger connection.

File: changelog.d/12766.bugfix [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Implement [MSC3816](https://github.com/matrix-org/matrix-spec-proposals/pull/3816): sending the root event in a thread 
+should count as "participated" in it.
```
Comment (erikjohnston) [inline changelog.d/12766.bugfix:2]:
```suggestion
Implement [MSC3816](https://github.com/matrix-org/matrix-spec-proposals/pull/3816): sending the root event in a thread should count as "participated" in it.
```

**Review by erikjohnston — APPROVED** [review]

File: synapse/handlers/relations.py (hunk 2 above)
Comment (clokep) [reply synapse/handlers/relations.py]: I clarified this and the surrounding comments in b88cac08976af7f49ccddc9a2a2583dbeff4cf8b. I think it is clearer now!

File: tests/rest/client/test_relations.py (hunk 3 above)
Comment (clokep) [reply tests/rest/client/test_relations.py:1081]: I attempted to clarify this a bit with comments in b88cac08976af7f49ccddc9a2a2583dbeff4cf8b.

**Comment by clokep** [comment]
Note that the attached MSC is now in FCP. 🎉
